src/resources/basis-worker.js

```diff
@@ -157,8 +157,98 @@ function BasisWorker() {
+        }
+
+        // choose the target format
+        const format = chooseTargetFormat(options.deviceDetails, hasAlpha, isUASTC);
+
+        // unswizzle gggr textures under pvr compression
+        const unswizzle = !!options.isGGGR && format === 'pvr';
+
+        // convert to basis format taking into consideration platform restrictions
+        let basisFormat;
+        if (unswizzle) {
+            // in order to unswizzle we need gggr8888
+            basisFormat = BASIS_FORMAT.cTFRGBA32;
+        } else {
+            // select output format based on supported formats
+            basisFormat = hasAlpha ? alphaMapping[format] : opaqueMapping[format];
+
+            // transcode to uncompressed format if the texture is PVRTC or webgl1, and has invalid dimensions
+            const isPVRTC = (basisFormat === BASIS_FORMAT.cTFPVRTC1_4_RGB || basisFormat === BASIS_FORMAT.cTFPVRTC1_4_RGBA);
+            const notPOT = ((width & (width - 1)) !== 0) || ((height & (height - 1)) !== 0);
+            const notSquare = (width !== height);
+            if ((isPVRTC && (notPOT || notSquare)) || (!options.deviceDetails.webgl2 && notPOT)) {
+                basisFormat = hasAlpha ? BASIS_FORMAT.cTFRGBA32 : BASIS_FORMAT.cTFRGB565;
+            }
+        }
+
+        if (!basisFile.startTranscoding()) {
+            basisFile.close();
+            basisFile.delete();
+            throw new Error('Failed to start transcoding url=' + url);
+        }
+
+        let i;
+
+        const levelData = [];
+        for (let mip = 0; mip < levels; ++mip) {
+            const dstSize = basisFile.getImageTranscodedSizeInBytes(mip, 0, 0, basisFormat);
+            const dst = new Uint8Array(dstSize);
+
+            if (!basisFile.transcodeImage(dst, mip, 0, 0, basisFormat, 0, -1, -1)) {
+                basisFile.close();
+                basisFile.delete();
+                throw new Error('Failed to transcode image url=' + url);
+            }
+
+            const is16BitFormat = (basisFormat === BASIS_FORMAT.cTFRGB565 || basisFormat === BASIS_FORMAT.cTFRGBA4444);
+
+            levelData.push(is16BitFormat ? new Uint16Array(dst.buffer) : dst);
+        }
+
+        basisFile.close();
+        basisFile.delete();
+
+        // handle unswizzle option
+        if (unswizzle) {
+            basisFormat = BASIS_FORMAT.cTFRGB565;
+            for (i = 0; i < levelData.length; ++i) {
+                levelData[i] = pack565(unswizzleGGGR(levelData[i]));
+            }
+        }
+
+        const compressedFormat = isCompressedFormat(basisFormat);
+
+        return {
+            format: basisToEngineMapping(basisFormat, options.deviceDetails),
+            width: compressedFormat ? ((width + 3) & ~3) : width,
+            height: compressedFormat ? ((height + 3) & ~3) : height,
+            levels: levelData,
+            cubemap: false,
+            mipmaps: true,
```

Comment:
Bit confused about this. So KTX2 won't contain non-mipmapped texture data?

Reply:
Ah yes nice catch. This mipmap flag is actually ignored (https://github.com/slimbuck/engine/blob/ktx2-support/src/resources/parser/texture/ktx2.js#L97) and shouldn't been here. I've removed it.